src/algebra/monoid_algebra/division.lean

```diff
@@ -0,0 +1,176 @@
+
+namespace add_monoid_algebra
+
+section
+variables [canonically_ordered_add_monoid G] [has_sub G] [has_ordered_sub G]
+  [contravariant_class G G (+) (≤)]
+
+/-- Divide by `of' k G g`, discarding terms not divisible by this. -/
+noncomputable def div_of (g : G) : add_monoid_algebra k G →+ add_monoid_algebra k G :=
+@finsupp.comap_domain.add_monoid_hom _ _ _ _ ((+) g)
+  (by letI := canonically_ordered_add_monoid.to_add_cancel_comm_monoid G;
+    exact add_right_injective g)
+
+@[simp] lemma div_of_apply (g : G) (x : add_monoid_algebra k G) (g' : G) :
+  div_of g x g' = x (g + g') := rfl
+
+@[simp] lemma support_div_of (g : G) (x : add_monoid_algebra k G)  :
+  (div_of g x).support = x.support.preimage ((+) g)
+    (function.injective.inj_on
+      (by letI := canonically_ordered_add_monoid.to_add_cancel_comm_monoid G;
+        exact add_right_injective g) _) := rfl
+
+lemma div_of_zero (x : add_monoid_algebra k G) : div_of 0 x = x :=
+by { ext, simp only [add_monoid_algebra.div_of_apply, zero_add] }
+
+lemma div_of_add (x : add_monoid_algebra k G) (a b : G) :
+  div_of (a + b) x = div_of b (div_of a x) :=
+by { ext, simp only [add_monoid_algebra.div_of_apply, add_assoc] }
+
+@[simp] lemma div_of_of'_mul (a : G) (x : add_monoid_algebra k G) :
+  div_of a (of' k G a * x) = x :=
+begin
+  letI := canonically_ordered_add_monoid.to_add_cancel_comm_monoid G,
+  ext b,
+  rw [add_monoid_algebra.div_of_apply, of'_apply, single_mul_apply_aux, one_mul],
+  intro c,
+  exact add_right_inj _,
+end
+
+@[simp] lemma div_of_mul_of' (a : G) (x : add_monoid_algebra k G) :
+  div_of a (x * of' k G a) = x :=
+begin
+  letI := canonically_ordered_add_monoid.to_add_cancel_comm_monoid G,
+  ext b,
+  rw [add_monoid_algebra.div_of_apply, of'_apply, mul_single_apply_aux, mul_one],
+  intro c,
+  rw add_comm,
+  exact add_right_inj _,
+end
+
+@[simp] lemma div_of_of' (a : G) : div_of a (of' k G a) = 1 :=
+by simpa only [one_mul] using div_of_mul_of' a (1 : add_monoid_algebra k G)
+
+/-- The remainder upon division by `of' k G g`. -/
+noncomputable def mod_of (g : G) (x : add_monoid_algebra k G) : add_monoid_algebra k G :=
+x.filter (λ g', ¬g ≤ g')
+
+@[simp] lemma mod_of_apply_of_not_le {g : G} (x : add_monoid_algebra k G) {g' : G} (h : ¬g ≤ g') :
+  mod_of g x g' = x g' :=
+finsupp.filter_apply_pos _ _ h
+
+@[simp] lemma mod_of_apply_of_le {g : G} (x : add_monoid_algebra k G) {g' : G} (h : g ≤ g') :
+  mod_of g x g' = 0 :=
+finsupp.filter_apply_neg _ _ (not_not_intro h)
+
+lemma single_mul_apply_of_not_le (r : k) {g g' : G} (x : add_monoid_algebra k G) (h : ¬(g ≤ g')):
+  (finsupp.single g r * x : add_monoid_algebra k G) g' = 0 :=
+begin
+  classical,
+  rw [mul_apply, finsupp.sum_single_index],
+  swap,
+  { simp_rw [finsupp.sum, zero_mul, if_t_t, finset.sum_const_zero] },
+  { apply finset.sum_eq_zero,
+    simp_rw ite_eq_right_iff,
+    rintros g'' hg'' rfl,
+    exfalso,
+    apply h,
+    apply le_add_of_le_of_nonneg le_rfl (zero_le g'') }
+end
```

Comment:
As you mentioned, this lemma and the one you flagged in #15983 are very similar.  I wonder whether there is a good choice of assumptions that makes both of our lemmas specialization of a common one.

Reply:
I can't prove your exact statement, but this result seems related (and I have a proof of it):
```lean
example {R A} [semiring R] [has_add A] [partial_order A] [covariant_class A A (+) (<)]
  [covariant_class A A (function.swap (+)) (<)] {a b : A} {f g : A →₀ R} 
  (hf : ∀ (i : A), i ∈ f.support → a ≤ i) (hg : ∀ (i : A), i ∈ g.support → b < i) :
  (f * g : add_monoid_algebra R A) (a + b) = 0 :=
```

Reply:
I checked: with the lemmas in my PR, I need to assume `h' : g' ≤ g` and `[covariant_class A A (+) (<)]` to get the lemma in yours.

Reply:
I ended up using `∃ d, d + g = g'` instead of an inequality, and now everything should hold much more generally.